### ISSUE TITLE
Deeplinkdispatch with Android Annotations - error creating file

### ISSUE THREAD

**Issue by somghosh** [excilys/androidannotations#1510]
I have a 1 activity project, with Android Annotations (using @EActivity to refer to the activity layout) and I am using DeeplinkDispatch on top of it

Seems like both annotations library don't go so well. If I remove either package, the app runs, but not with both

## Only activity class

	package com.example.android.deeplinktest;

	import android.support.v7.app.AppCompatActivity;
	import android.os.Bundle;
	import android.view.Menu;
	import android.view.MenuItem;
	import android.widget.TextView;

	import com.airbnb.deeplinkdispatch.DeepLink;

	import org.androidannotations.annotations.AfterViews;
	import org.androidannotations.annotations.EActivity;

	@EActivity(R.layout.activity_main)
	@DeepLink("foo://deepLink/{id}")
	public class MainActivity extends AppCompatActivity {

	/*    @Override
	    protected void onCreate(Bundle savedInstanceState) {
	        super.onCreate(savedInstanceState);
	        setContentView(R.layout.activity_main);

	    }*/

	    @AfterViews
	    public void init(){
	        if (getIntent().getBooleanExtra(DeepLink.IS_DEEP_LINK, false)) {
	            Bundle parameters = getIntent().getExtras();

	            String idString = parameters.getString("id");

	            ((TextView)findViewById(R.id.tv)).setText(idString);
	        }
	    }

	}

## Android Manifest

	<?xml version="1.0" encoding="utf-8"?>
	<manifest xmlns:android="http://schemas.android.com/apk/res/android"
	    package="com.example.android.deeplinktest" >

	    <application
	        android:allowBackup="true"
	        android:icon="@mipmap/ic_launcher"
	        android:label="@string/app_name"
	        android:theme="@style/AppTheme" >
	        <activity
	            android:name=".MainActivity_"
	            android:label="@string/app_name" >
	            <intent-filter>
	                <action android:name="android.intent.action.MAIN" />

	                <category android:name="android.intent.category.LAUNCHER" />
	            </intent-filter>
	        </activity>

	        <activity
	            android:name="com.airbnb.deeplinkdispatch.DeepLinkActivity"
	            android:theme="@android:style/Theme.NoDisplay">
	            <intent-filter>
	                <action android:name="android.intent.action.VIEW" />
	                <category android:name="android.intent.category.DEFAULT" />
	                <category android:name="android.intent.category.BROWSABLE" />
	                <data android:scheme="foo" />
	            </intent-filter>
	        </activity>

	    </application>

	</manifest>

## build.gradle

	apply plugin: 'com.android.application'
	apply plugin: 'com.neenbedankt.android-apt'

	android {
	    compileSdkVersion 22
	    buildToolsVersion "22.0.1"

	    defaultConfig {
	        applicationId "com.example.android.deeplinktest"
	        minSdkVersion 17
	        targetSdkVersion 22
	        versionCode 1
	        versionName "1.0"
	    }
	    buildTypes {
	        release {
	            minifyEnabled false
	            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
	        }
	    }
	}

	apt {
	    arguments {
	        androidManifestFile variant.outputs[0].processResources.manifestFile
	        resourcePackageName "com.example.android.deeplinktest"
	    }
	}

	dependencies {
	    compile fileTree(dir: 'libs', include: ['*.jar'])
	    compile 'com.android.support:appcompat-v7:22.2.1'
	    compile 'com.airbnb:deeplinkdispatch:1.3.0'
	    apt 'com.airbnb:deeplinkdispatch-processor:1.3.0'

	    apt "org.androidannotations:androidannotations:3.2"
	    compile 'org.androidannotations:androidannotations-api:3.2'

	}

## Error message

	Information:Gradle tasks [clean, :app:compileDebugSources, :app:compileDebugAndroidTestSources]
	:app:clean
	:app:preBuild UP-TO-DATE
	:app:preDebugBuild UP-TO-DATE
	:app:checkDebugManifest
	:app:preReleaseBuild UP-TO-DATE
	:app:prepareComAndroidSupportAppcompatV72221Library
	:app:prepareComAndroidSupportSupportV42221Library
	:app:prepareDebugDependencies
	:app:compileDebugAidl
	:app:compileDebugRenderscript
	:app:generateDebugBuildConfig
	:app:generateDebugAssets UP-TO-DATE
	:app:mergeDebugAssets
	:app:generateDebugResValues UP-TO-DATE
	:app:generateDebugResources
	:app:mergeDebugResources
	:app:processDebugManifest
	:app:processDebugResources
	:app:generateDebugSources
	:app:processDebugJavaRes UP-TO-DATE
	:app:compileDebugJava
	Note: Resolve log file to /Users/somghosh/Programming/DeepLinkTest/app/build/generated/source/apt/androidannotations.log
	Note: Initialize AndroidAnnotations 3.2 with options {androidManifestFile=/Users/somghosh/Programming/DeepLinkTest/app/build/intermediates/manifests/full/debug/AndroidManifest.xml, resourcePackageName=com.example.android.deeplinktest}
	Note: Start processing for 2 annotations on 4 elements
	Note: AndroidManifest.xml file found with specified path: /Users/somghosh/Programming/DeepLinkTest/app/build/intermediates/manifests/full/debug/AndroidManifest.xml
	Warning:A class activity declared in the AndroidManifest.xml cannot be found in the compile path: [com.airbnb.deeplinkdispatch.DeepLinkActivity]
	Note: AndroidManifest.xml found: AndroidManifest [applicationPackage=com.example.android.deeplinktest, componentQualifiedNames=[com.example.android.deeplinktest.MainActivity_], permissionQualifiedNames=[], applicationClassName=null, libraryProject=false, debugabble=false, minSdkVersion=17, maxSdkVersion=-1, targetSdkVersion=22]
	Note: Found project R class: com.example.android.deeplinktest.R
	Note: Found Android class: android.R
	Note: Validating elements
	Note: Validating with EActivityHandler: [com.example.android.deeplinktest.MainActivity]
	Note: Validating with AfterViewsHandler: [init()]
	Note: Processing root elements
	Note: Processing root elements EActivityHandler: [com.example.android.deeplinktest.MainActivity]
	Note: Processing enclosed elements
	Note: Number of files generated by AndroidAnnotations: 1
	Note: Writting following API classes in project: []
	Note: Generating class: com.example.android.deeplinktest.MainActivity_
	Note: Time measurements: [Whole Processing = 259 ms], [Extract Manifest = 178 ms], [Process Annotations = 32 ms], [Generate Sources = 18 ms], [Find R Classes = 14 ms], [Extract Annotations = 4 ms], [Validate Annotations = 4 ms], 
	Note: Finish processing
	Error:Error creating file
	Note: Start processing for 0 annotations on 3 elements
	Note: Time measurements: [Whole Processing = 0 ms], 
	Note: Finish processing
	Note: Start processing for 0 annotations on 0 elements
	Note: Time measurements: [Whole Processing = 1 ms], 
	Note: Finish processing
	1 warning

**Comment by dodgex** [comment]
for now i assume an issue in DeepLinkDispatch see as the error comes from its processor: https://github.com/airbnb/DeepLinkDispatch/issues/38#issuecomment-127935109

**Comment by andrewlord1990** [comment]
The error is an IOException when trying to write the Deep link dispatch classes to file. Is there a chance Android Annotations does something whilst it is generated classes that would prevent DeepLinkDispatch to also create a class before Android Annotations has finished?

**Comment by WonderCsabo** [comment]
Guys, please provide a minimal example project.

**Comment by WonderCsabo** [comment]
Closing this issues due to lack of response. Please reopen if you still have problems.

**Comment by felipecsl** [comment]
Reposting from the DeepLinkDispatch issue. Here is the relevant error message I'm seeing in the build output, could be the reason for the problem:

```
error: Could not find the AndroidManifest.xml file, going up from path [/Users/felipe_lima/prj/DeepLinkDispatch/sample/build/generated/source/apt/debug] found using dummy file [] (max atempts: file:///Users/felipe_lima/prj/DeepLinkDispatch/sample/build/generated/source/apt/debug/dummy1442168401108)
```

**Comment by WonderCsabo** [comment]
That is indeed a problem. You did not configure AA correctly. You have to tell AA where can it find the manifest file, as written [here](https://github.com/excilys/androidannotations/wiki/Building-Project-Gradle). Add this to your *build.gradle*:

```groovy
apt {
    arguments {
        androidManifestFile variant.outputs[0].processResources.manifestFile
}
```

**Comment by dodgex** [comment]
there is one `}` missing

**Comment by WonderCsabo** [comment]
Fixed, thanks!

**Comment by felipecsl** [comment]
@WonderCsabo that didnt quite work. Had to use this instead:

```
apt {
    arguments {
        androidManifestFile android.applicationVariants.outputs[0].processResources.manifestFile
    }
}
```

However, still same problem. The error message is slightly different now though:

```
Note: Resolve log file to /Users/felipe_lima/prj/DeepLinkDispatch/sample/build/generated/source/apt/androidannotations.log
Note: Initialize AndroidAnnotations 3.3.2 with options {androidManifestFile=[/Users/felipe_lima/prj/DeepLinkDispatch/sample/build/intermediates/manifests/full/debug/AndroidManifest.xml]}
Note: Start processing for 1 annotations on 6 elements
error: Could not find the AndroidManifest.xml file in specified path : [/Users/felipe_lima/prj/DeepLinkDispatch/sample/build/intermediates/manifests/full/debug/AndroidManifest.xml]
Note: Time measurements: [Whole Processing = 13 ms], [Extract Manifest = 10 ms], [Extract Annotations = 2 ms],
Note: Finish processing
Note: Start processing for 0 annotations on 0 elements
Note: Time measurements: [Whole Processing = 4 ms],
Note: Finish processing
```

I checked this path and the file does exist: `/Users/felipe_lima/prj/DeepLinkDispatch/sample/build/intermediates/manifests/full/debug/AndroidManifest.xml`.

Anything I'm missing here?

**Comment by WonderCsabo** [comment]
This is very weird. Can you provide us the project, or a minimal reproducing example project?

**Comment by felipecsl** [comment]
There you go, it's in the `issue/android-annotations` branch: https://github.com/airbnb/DeepLinkDispatch/tree/issue/android-annotations/sample

**Comment by dodgex** [comment]
can you provide a project that actually builds? i get a failing test

**Comment by dodgex** [comment]
and: 
```
:sample:compileDebugJavaWithJavac
\DeepLinkDispatch\sample\src\main\java\com\airbnb\deeplinkdispatch\sample\DeepLinkReceiver.java:8: error: cannot find symbol
import com.airbnb.deeplinkdispatch.DeepLinkActivity;
                                  ^
  symbol:   class DeepLinkActivity
  location: package com.airbnb.deeplinkdispatch
```

**Comment by felipecsl** [comment]
you can ignore the failing test. The compilation error happens exactly because of the AndroidAnnotations issue. If you remove the `@EActivity(R.layout.activity_main)` annotation from `MainActivity`, it builds fine

**Comment by dodgex** [comment]
there is no `DeepLinkActivity` in the jars? is that the generated class?

**Comment by dodgex** [comment]
if yes, the problem in that sample seems to be that AA is run before DeepLinkDispatch so that the class is not yet generated. the rror is not from AA but from `javac`

**Comment by felipecsl** [comment]
yep, is there a way to force order of execution with apt?

**Comment by dodgex** [comment]
no idea. maybe @hvisser can answer that.

**Comment by WonderCsabo** [comment]
You can specify the processors manually. In that case Javac should call the processors in the order you defined them:
```groovy
apt {
    arguments {
        androidManifestFile android.applicationVariants.outputs[0].processResources.manifestFile
    }
    
    processor "com.airbnb.deeplinkdispatch.DeepLinkProcessor"
    processor "org.androidannotations.AndroidAnnotationsProcessor"

    disableDiscovery true
}
˙˙˙

**Comment by felipecsl** [comment]
didn't work either: `error: Annotation processor 'org.androidannotations.AndroidAnnotationsProcessor' not found`. I can't find neither processor

**Comment by WonderCsabo** [comment]
Sorry... org.androidannotations.AndroidAnnotationProcessor is the class.

https://github.com/excilys/androidannotations/blob/master/AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/AndroidAnnotationProcessor.java

**Comment by felipecsl** [comment]
doesnt work either

**Comment by felipecsl** [comment]
okay removing the `disableDiscovery` option worked. Here's the final setup:

```

**Comment by WonderCsabo** [comment]
@felipecsl are you sure? Processor call order is random, so you may have to test it multiple times to assert the correct order.

@hvisser can you clarify this for us? `disableDiscovery` should not result in processors not found when specified with `processor`. :S

**Comment by felipecsl** [comment]
yep, I can reproduce 100% of the time

**Comment by felipecsl** [comment]
ugh.. never mind, still the same problem. I had the `@EActivity` annotation commented out

**Comment by felipecsl** [comment]
still fails with the `Could not find the AndroidManifest.xml` error. It's just that apt can find the processors now

**Comment by hvisser** [comment]
`disableDiscovery` simply does what it says: it will only run the annotation processors specified by `processor` *if* they are found on the class path. This means that these processors probably should not be in the `apt` scope (but I'll have to double check). 
@felipecsl If you know the path to the manifest that is generated, you can also just hardcode it, at least for debugging. I've added some more examples to the `android-apt` README: you can also add code to the arguments section to do whatever you want to set an argument.

**Comment by hvisser** [comment]
Oh by the way, uses reported to me that since the 1.3.0 and `android-apt` 1.7 version of the Android Gradle plugin this should work for AA:

`androidManifestFile variant.outputs[0]?.processResources?.manifestFile`

Which is more bullet proof for variants (e.g. tests) that don't have manifest files.

**Comment by ray1984** [comment]
What about? Is it real to fix?

**Comment by WonderCsabo** [comment]
Actually i do not know... Actually the problem @felipecsl sees is totally different from the first post. It would be nice if you would share your projects so we can investigate.

**Comment by WonderCsabo** [comment]
@hvisser Actually it is possible to define the annotation prcessor order. They will be called in the order what you specify in the `-processor` flag in `javac`. In `android-apt`, this can be done in the following way:
```groovy
dependencies {
    compile "org.androidannotations:androidannotations-api:3.3.2"
    provided "org.androidannotations:androidannotations:3.3.2"

    compile "com.squareup.dagger:dagger:1.2.2"
    provided "com.squareup.dagger:dagger-compiler:1.2.2" // apt will not work!
}

apt {
    processor "org.androidannotations.AndroidAnnotationProcessor"

    processor "dagger.internal.codegen.ValidationProcessor"
    processor "dagger.internal.codegen.InjectAdapterProcessor"
    processor "dagger.internal.codegen.ModuleAdapterProcessor"
    processor "dagger.internal.codegen.GraphAnalysisProcessor"

    disableDiscovery true
}
```
I verified that the processors will run in the specified order, i also switched the order.

The only drawback of this method due to the `provided` scope, processor dependencies will show up in the IDE external dependency section, so you can accidentally reference them in the code.
But if we use `apt` scope, `javac` will tell that he cannot find the processor class.
Is there a way to overcome this problem (use this manual processor class specification, but still use some `apt` scope so the processor will not end in in the IDE as compile dependencies)? Thanks in advance!

**Comment by hvisser** [comment]
That's just an implementation detail of javac, not something that you can
rely on. This (older) doc describes pretty well how processing takes place:
http://docs.oracle.com/javase/7/docs/technotes/tools/solaris/javac.html

Especially: _After scanning the source files and classes on the command
line to determine what annotations are present, the compiler queries the
processors to determine what annotations they process. When a match is
found, the processor will be invoked. A processor may "claim" the
annotations it processes, in which case no further attempt is made to find
any processors for those annotations. Once all annotations have been
claimed, the compiler does not look for additional processors.

If any processors generate any new source files, another round of
annotation processing will occur: any newly generated source files will be
scanned, and the annotations processed as before. Any processors invoked on
previous rounds will also be invoked on all subsequent rounds. This
continues until no new source files are generated._

**Comment by WonderCsabo** [comment]
@hvisser i see, thanks for the detailed answer. We cannot rely on it, but some cases it is needed, so let's called this a hack. Unfortunately in some cases, we have to define the order, at least of the first round. Of course, processors should be able to handle error types, but that is far from trivial to implement.

**Comment by hvisser** [comment]
You shouldn't need the `processor` options at all if the processors are declared in the services files. I don't know why you need `provided` for `dagger-compiler`, it doesn't make sense unless there's stuff in dagger-compiler` that needs to be on your class path. When you set `disableDiscovery` the `processorPath` of `javac` is not set, and effectively `apt` scope is disabled.

**Comment by hvisser** [comment]
@WonderCsabo Yeah, that would be your hack biting you in the ass in that case ;)

**Comment by WonderCsabo** [comment]
I only used `provided` because as you said, `apt` scope is disabled.

Hmm, i think you are right. Maybe setting the processor order does not help at all. Even if the first annotation processor generates a class, which is referenced in an annotation claimed by the second processor, the type would be still an error type, because in the same round, generated classes are not yet added to the Java model. I guess they are only available in the second round.

**Comment by iHehaDev** [comment]
is this error still happen in the current version ?

```
apt {
    arguments {
        androidManifestFile variant.outputs[0]?.processResources?.manifestFile
    }

    processor "org.androidannotations.AndroidAnnotationProcessor"
    processor "com.airbnb.deeplinkdispatch.DeepLinkProcessor"

    disableDiscovery true
}
```

will produce error of the below. 

```
Error:Annotation processor 'org.androidannotations.AndroidAnnotationProcessor' not found
```

**Comment by WonderCsabo** [comment]
What do you mean by the *current version*?

In 4.0-SNAPSHOT, the processor class is `org.androidannotations.internal.AndroidAnnotationProcessor`

**Comment by iHehaDev** [comment]
Sorry that I make you confused. I am using 3.3.2 which is in document so I can assume it is stable release ?

This one only fix in 4.0-SNAPSHOT ?

**Comment by WonderCsabo** [comment]
Yeah, the current stable is 3.3.2. I do not know why it cannot find the processor class. 

There was no fix related to this issue.

**Comment by iHehaDev** [comment]
@WonderCsabo any recommendation if our team want to use androidannotations and Deeplinkdispatch together ? Thanks for your help.

**Comment by WonderCsabo** [comment]
@iHehaDev @felipecsl i think i found what is the problem.

There are two annotation processing rounds in the [sample](https://github.com/airbnb/DeepLinkDispatch/tree/issue/android-annotations/sample) project:

1. round: `MainActivity` is processed
2. round: `MainActivity_` is processed

In the second round `MainActivity_` is processed, because AA adds the `@DeepLink` annotation to the generated `MainActivity_` class. This is actually needed, because the `DeepLingActivity` should start `MainActivity_` and not `MainActivity`.

The error is raised when deep link processor tries to generate the `DeepLinkLoader.java` file in the 2. round. It was already generated in the 1. round, hence the .java file cannot be created, and the processing fails with an exception.

TL;DR: `DeepLinkProcessor` should be prepared to deal with subsequent annotation processing rounds. Currently if it is called in the subsequent round and `@DeepLink` annotation is added in the previous round (by any processor), it will fail.

**Comment by hvisser** [comment]
...Or AA shouldn't copy that annotation in the first place. It's questionable if it should do so I think. I know nothing about the `DeepLinkProcessor` here, but if it did generate code for the `MainActivity_` class, would that even work? User code would still only know about the one generated for `MainActivity` I guess.

**Comment by WonderCsabo** [comment]
@hvisser we copy annotations to our generated classes because they are actually "replacing" the original ones.

Well, i do not know actually it could work, but i know that if we do not copy the annotation it will definitely not work. The generated `DeepLinkActivity` should start the generated `MainActivity_` class, and not the original `MainActivity` class.

**Comment by iHehaDev** [comment]
@WonderCsabo Agree with you, the deep link should process the ```MainActivity_``` as it should be the Activity that the url hook to.

**Comment by felipecsl** [comment]
`DeepLinkProcessor` is using `Element.toString()` to get the full qualified class for the annotated class, so that should be `MainActivity_` if that's the class with the annotation. I don't see a problem here.
Also I can't seem to be able to make AA work on that sample branch. I keep getting 
```
error: Could not find the AndroidManifest.xml file in specified path : [/Users/felipe_lima/prj/DeepLinkDispatch/sample/build/intermediates/manifests/full/debug/AndroidManifest.xml]
```
when I use
```
apt {
    arguments {
        androidManifestFile android.applicationVariants.outputs[0].processResources.manifestFile
    }
}
```
Even if I hardcode the `androidManifestFile` to `/Users/felipe_lima/prj/DeepLinkDispatch/sample/build/intermediates/manifests/full/debug/` I still get this error:
```
error: Could not parse the AndroidManifest.xml file at path /Users/felipe_lima/prj/DeepLinkDispatch/sample/build/intermediates/manifests/full/debug
```

**Comment by WonderCsabo** [comment]
@felipecsl there is no problem with the name of the `Activity`. The problem lies [here](https://github.com/airbnb/DeepLinkDispatch/blob/master/deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java#L105). In the second annotation processing round, the processor tries to create a file (*DeepLinkLoader.java*). But that file was already created in the first round, hence an exception is thrown when try to create it again.

I created a [standalone project](https://github.com/WonderCsabo/DeepLinkDispatchAA) which does not produce the `androidManifestFile` problem. You can see the error coming from the deepling processor (`error: Error creating file`) by running `./gradlew assembleDebug`.

**Comment by dodgex** [comment]
This is exactly what i already posten [7 month ago](https://github.com/airbnb/DeepLinkDispatch/issues/38#issuecomment-127935109) on the related issue on DeepLinkDispatch repo. Beside the fact that i didn't realize that it was about the second round.

**Comment by andrewlord1990** [comment]
If you have an activity, say `HomeActivity` there would then also be a `HomeActivity_`. Both of these activities would have the `@DeepLink` annotation on them and would both be picked up by the `DeepLinkProcessor`.

Therefore, firstly the `process` method would need to handle the situation of two processing rounds occurring. Then, it would also need to handle both `HomeActivity` and `HomeActivity_` responding to the same URL. We would need a way to ensure that `HomeActivity_` is the one that is launched. 

Any suggestions on how to ensure this? @felipecsl or @WonderCsabo. Not entirely sure where something like this should go, I am guessing some kind of generic solution in `DeepLinkDispatch` for an `Activity` extending another and then responding to same URL.

It seems that the problem is due to the generated class extending the regular one. We would need some way of telling `DeepLinkProcessor` to not store a reference in the registry for the non-underscored version of the class.

**Comment by WonderCsabo** [comment]
Yeah, a general solution is needed in `DeepLinkDispatch`, we cannot really control this from AA-side.

Unfortunately making annotation processor play well with each other is not a trivial task.

**Comment by iHehaDev** [comment]
I am thinking is it possible and proper to call DeepLink Processor within the AA, so that injection only on  HomeActivity_ but not on HomeActivity. 

But the question would be is it possible to create/update another file (DeepLinkActivity.java and DeepLinkLoader.java) out of the one with annotation written to it. 

@WonderCsabo how do you think ? any existing 3rd party do similar way ? 

somehow like below integration would be great if all file deeplink need can be generate after AA files generated. 
https://github.com/excilys/androidannotations/wiki/RoboGuiceIntegration

**Comment by iHehaDev** [comment]
I will try to do so if I can refer to existing library to build a new one and it would be great for both library user.

**Comment by alikian** [comment]
I experience same issue when try to integrate AA with MapStruct or Selma.

**Comment by ilyaklyukin** [comment]
here is my workaround for using AA + DeepLinkispatcher
https://github.com/airbnb/DeepLinkDispatch/issues/218#issuecomment-414350925

**Comment by Tecmax** [comment]
Hi Team, Any Solutions would be appreciated.
I get this unique error after migration to androidx 

this is in gradle
 implementation "com.airbnb:deeplinkdispatch:4.1.0"
    annotationProcessor "com.airbnb:deeplinkdispatch-processor:4.1.0"

and this is the appcompat lib as well
implementation 'androidx.appcompat:appcompat:1.0.0'

Any upgradations required or we have stopped androidx support


java.lang.RuntimeException: Unable to instantiate activity ComponentInfo{com.maersk.trackandtrace.maerskline.preprod/com.airbnb.deeplinkdispatch.DeepLinkActivity}: java.lang.ClassNotFoundException: Didn't find class "com.airbnb.deeplinkdispatch.DeepLinkActivity" on path: DexPathList[[zip file "/data/app/

**Comment by WonderCsabo** [comment]
@Tecmax this error is not related to AndroidAnnotations, sorry.
